### PR TITLE
#100 Implemented unsuccessful tap UI

### LINKED ISSUE #100: Change 'Collected' text to 'Broken Lens' when tap errors
Currently the 'Collect' element displays 'Collected ✔️' when the tap errors, this is confusing to the visitor.

Instead we want to show a 'Broken Lens X' message when the tap errors.

### PR DESCRIPTION
**Pull request by sam-maher-acmi**
Resolves #100 

When a tap fails to be created in XOS, an error dialogue now displays.

### Acceptance Criteria
- [x] Failure dialogue appears on failed tap
- [x] Dialogue matches design files

### Relevant design files
* https://trello.com/c/Y2He78ze/142-fix-playlist-label-interactive-label-and-spotlights-to-return-an-error-message-instead-of-collected-when-tapping-with-a-broken-l

### Testing instructions
1. Modify your config.env
   - XOS_API_ENDPOINT=https://xos.acmi.net.au/api/
   - XOS_TAPS_ENDPOINT=http://your-ip-address:5000/api/taps/
   - PLAYLIST_ID=44
2. Run interactive-label i.e. `cd development && docker-compose up`
3. Run your development lens reader on [e__taps-reader-pi-4](https://dashboard.balena-cloud.com/devices/81361d59c71358c33710191666890fe8/envvars) and modify it's environment variable to point to your spotlights
   - TARGET_API_ENDPOINT=http://your-ip-address:8081/api/
4. Run your favourite flask tap server! (or modify the one I used below with your own valid lens uids).
5. Open up your browser at localhost:8081, set the dimensions of the browser to 1920x1080 in the web console
5. Tap with a valid lens, notice the normal 'collected' UI response occurs
6. Tap with an invalid lens, notice the new error dialogue

### Notes
1. I edited the flask server to emulate invalid lenses. Feel free to do so as well (replacing the 'valid-lens-uid'), or of course use your own method.
```
# server.py
from flask import Flask, Response, request
app = Flask(__name__)

@app.route('/')
def home():
    return taps()

@app.route('/api/taps/', methods=['POST'])
def taps():
    response_json = request.get_json()
    print(response_json)
    lens_uid = response_json['lens']['uid']
    status_code = 201 if lens_uid in ['<valid-lens-uid>'] else 400
    return Response('{"ok": "true"}', status=status_code, mimetype='application/json')
```

### DoD
- [ ] All acceptance criteria are met
- [ ] New logic has been documented
- [ ] New logic has appropriate unit tests
- [ ] Changelog has been updated if necessary
- [ ] Deployment / migration instruction have been updated if required

### CODE REVIEW

File: app/main.py [hunk 1]
```diff
@@ -149,20 +150,28 @@ def collect_item():
...
+    if not has_tapped:
+        return HTTPError('Tap still processing.'), abort(500, 'Tap still processing.')
+    has_tapped.tap_successful = 0
+    has_tapped.has_tapped = 1
+
     xos_tap = dict(request.get_json())
     try:
         record = model_to_dict(Label.select().order_by(Label.datetime.desc()).get())
     except DoesNotExist:
+        has_tapped.save()
```
Comment (sighmon) [inline app/main.py:162]: Separating the `.save()` from where you set the data doesn't feel right. What was your logic behind doing this?

File: app/static/app.js [hunk 2]
```diff
@@ -394,25 +395,39 @@ function open_tap_error() {
 }
 
 const tap_source = new EventSource("/api/tap-source");
-tap_source.onmessage = function () {
+
+tap_source.onmessage = function (e) {
```
Comment (sighmon) [inline app/static/app.js:399]:
```suggestion
tap_source.onmessage = function (message) {
```

Single character variables are discouraged... not sure how the linter didn't catch this one.

So maybe call it `message` or `event`.

File: app/main.py [hunk 3]
```diff
@@ -188,5 +197,5 @@ def tap_source():
 
 if __name__ == '__main__':
     db.create_tables([Label, HasTapped])
-    HasTapped.create(has_tapped=0)
+    HasTapped.create(has_tapped=0, tap_successful=0)
```
Comment (sighmon) [inline app/main.py:200]: I suspect that `label.db` is deleted when the Interactive Label starts, but you might want to handle the following error incase it isn't. Else we'll need to manually delete `label.db` on all the interactive labels when they update to this code.

```bash
labelinteractive    | Traceback (most recent call last):
labelinteractive    |   File "/usr/local/lib/python3.7/runpy.py", line 193, in _run_module_as_main
labelinteractive    |     "__main__", mod_spec)
labelinteractive    |   File "/usr/local/lib/python3.7/runpy.py", line 85, in _run_code
labelinteractive    |     exec(code, run_globals)
labelinteractive    |   File "/code/app/main.py", line 200, in <module>
labelinteractive    |     HasTapped.create(has_tapped=0, tap_successful=0)
labelinteractive    |   File "/usr/local/lib/python3.7/site-packages/peewee.py", line 6346, in create
labelinteractive    |     inst.save(force_insert=True)
labelinteractive    |   File "/usr/local/lib/python3.7/site-packages/peewee.py", line 6556, in save
labelinteractive    |     pk = self.insert(**field_dict).execute()
labelinteractive    |   File "/usr/local/lib/python3.7/site-packages/peewee.py", line 1907, in inner
labelinteractive    |     return method(self, database, *args, **kwargs)
labelinteractive    |   File "/usr/local/lib/python3.7/site-packages/peewee.py", line 1978, in execute
labelinteractive    |     return self._execute(database)
labelinteractive    |   File "/usr/local/lib/python3.7/site-packages/peewee.py", line 2745, in _execute
labelinteractive    |     return super(Insert, self)._execute(database)
labelinteractive    |   File "/usr/local/lib/python3.7/site-packages/peewee.py", line 2474, in _execute
labelinteractive    |     cursor = database.execute(self)
labelinteractive    |   File "/usr/local/lib/python3.7/site-packages/peewee.py", line 3157, in execute
labelinteractive    |     return self.execute_sql(sql, params, commit=commit)
labelinteractive    |   File "/usr/local/lib/python3.7/site-packages/peewee.py", line 3151, in execute_sql
labelinteractive    |     self.commit()
labelinteractive    |   File "/usr/local/lib/python3.7/site-packages/peewee.py", line 2917, in __exit__
labelinteractive    |     reraise(new_type, new_type(exc_value, *exc_args), traceback)
labelinteractive    |   File "/usr/local/lib/python3.7/site-packages/peewee.py", line 190, in reraise
labelinteractive    |     raise value.with_traceback(tb)
labelinteractive    |   File "/usr/local/lib/python3.7/site-packages/peewee.py", line 3144, in execute_sql
labelinteractive    |     cursor.execute(sql, params or ())
labelinteractive    | peewee.OperationalError: table hastapped has no column named tap_successful
```

File: app/main.py [hunk 4]
```diff
@@ -149,20 +150,28 @@ def collect_item():
     Collect a tap and forward it on to XOS with the label ID.
     """
     has_tapped = HasTapped.get_or_none(has_tapped=0)
-    if has_tapped:
-        has_tapped.has_tapped = 1
-        has_tapped.save()
+    if not has_tapped:
+        return HTTPError('Tap still processing.'), abort(500, 'Tap still processing.')
```
Comment (sighmon) [inline app/main.py:154]: I don't think we should return a `500` error here, else the Lens Reader will re-queue the tap and try to send it to the Interactive Label again.

How about we either implement Tap queueing or ignore taps while we're processing previous ones?

**Review by sighmon — CHANGES_REQUESTED** [review]
Let's have a chat about this one tomorrow and talk through some of the changes I've suggested. A couple are questions I'm not totally sure about.

File: app/main.py (hunk 3 above)
Comment (sam-maher-acmi) [reply app/main.py:200]: Quickly tested `db.drop_tables([Label, HasTapped])` out, which seems to work.

But as discussed, will leave it as is for now.

File: app/main.py (hunk 4 above)
Comment (sam-maher-acmi) [reply app/main.py]: As discussed, implemented `tap_processing` to be explicit about this logic, and avoid unwanted requeuing of taps.

Let me know what you think.

File: app/static/app.js (hunk 2 above)
Comment (sam-maher-acmi) [reply app/static/app.js]: Done.

File: app/main.py (hunk 1 above)
Comment (sam-maher-acmi) [reply app/main.py]: As discussed, explicitly set data where saving. Along with `tap_processing` the logic is all a bit more readable now.

**Comment by sighmon** [comment]
@sam-maher-acmi I've got this setup:
* Lens Reader pointing to Interactive Label on port 8081
* Interactive Label pointing to XOS on my dev laptop 8000

Because it's all running on the same machine, when I tap I regularly hit the timeout on the Lens Reader because the Interactive Label doesn't respond in time. 

The response from XOS is successful eventually, but because the Lens Reader gets the timeout, it re-queues the tap to try and send again to the Interactive Label.

This results in multiple taps being created, one for each timeout failure.

So, I'm thinking we should either:
* Reply straight after we try and retrieve the Label, and have the only failure UI on the Interactive Label
* Send the Tap back to the Lens Reader, and implement a failure API to trigger the UI failure message on the Interactive Label

What do you think? This might need some UX design help from Pip/Lucie.

**Comment by sam-maher-acmi** [comment]
@sighmon 
As discussed, appears the time-out issue was pre-existing. Chucking this back into review.

Will create an issue/trello card for the time-out issue in any case.

**Review by sighmon — APPROVED** [review]
Looks good, thanks for the changes.

**Pull request merged**